keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -687,20 +997,92 @@ We will extend this approach:
 Note that this does allow `ValidatingAdmissionPolicy` to match
 `ValidatingWebhookConfiguration`.
 
-We are also considering not allowing a `ValidatingAdmissionPolicy` to match the
-configuration CRD kind that is uses for config. It would be possible to recover
-from this since the `ValidatingAdmissionPolicy` would remain accessible, but
-might require a privileged user do the repair.
-
-TODO: We should consider loosening this up and allow admission configuration to
-intercept/guard writes to admission configuration while preventing deadlock -
-Add feature to configure a set of webhooks to intercept other webhooks
-https://github.com/kubernetes/kubernetes/issues/101794.
+Note: In the future we may further loosen this up and allow admission
+configuration to intercept/guard writes to admission configuration while
+preventing deadlock - Add feature to configure a set of webhooks to intercept
+other webhooks https://github.com/kubernetes/kubernetes/issues/101794.
 
 Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
 `ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
 higher level. This could be added later.
 
+#### Reporting violations
+
+Goals:
+
+- Feature parity with `AdmissionReview`
+- Ability to format message strings
+- support for reasons/codes
+
+Example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  ...
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      level: Deny # the default
+      messageExpression: "self.name + ' must start with \"xyz-\"'"
+      reason: Invalid
+    - expression: "self.name.contains('bad')"
+      level: Warn
+      message: "name contains 'bad' which is discouraged due to ..."
+    - expression: "self.name.contains('suspicious')"
+      level: Audit
```

Comment:
Setting the enforcement level in the policy definition potentially makes it impossible for an admin to customize the level (unless the definition author parameterizes it). This makes it more challenging for an admin to have different severities in different scopes enforced via match criteria.

Use-case example: gradually rolling out a new policy, switching it from `Warn` to `Deny` namespace-by-namespace as pre-existing violations are resolved.

Reply:
I wonder if the both policy and the configurations should have some control of enforcement level? See my comments in above thread.

Reply:
Moved enforcement control entirely into policy configuration.